src/OrchardCore/OrchardCore.Queries.Core/Services/DefaultQueryManager.cs

```diff
@@ -0,0 +1,285 @@
+        {
+            await LoadAsync(record);
+        }
+
+        return records;
+    }
+
+    public async Task<Query> NewAsync(string source, JsonNode data = null)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(source);
+
+        var querySource = _serviceProvider.GetKeyedService<IQuerySource>(source);
+
+        if (querySource == null)
+        {
+            return null;
+        }
+
+        var query = new Query()
+        {
+            Source = source,
+        };
+
+        var initializingContext = new InitializingQueryContext(query, data);
+
+        await _queryHandlers.InvokeAsync((handler, context) => handler.InitializingAsync(context), initializingContext, _logger);
+
+        if (data != null)
+        {
+            // During creating a query, we attempt to set the query name.
+            query.Name = data[nameof(Query.Name)]?.GetValue<string>();
+
+            UpdateBaseProperties(data, query);
+        }
+
+        var initializedContext = new InitializedQueryContext(query);
+
+        await _queryHandlers.InvokeAsync((handler, context) => handler.InitializedAsync(context), initializedContext, _logger);
+
+        // Set the source again after calling handlers to prevent handlers from updating the source during initialization.
+        query.Source = source;
+
+        return query;
+    }
+
+    public async Task<ListQueryResult> PageQueriesAsync(int page, int pageSize, QueryContext context = null)
+    {
+        var query = GetQuery(context);
+
+        var skip = (page - 1) * pageSize;
+
+        var result = new ListQueryResult
+        {
+            Count = await query.CountAsync(),
+            Records = await query.Skip(skip).Take(pageSize).ListAsync()
+        };
+
+        foreach (var record in result.Records)
+        {
+            await LoadAsync(record);
+        }
+
+        return result;
+    }
+
+    public async Task UpdateAsync(Query query, JsonNode data = null)
+    {
+        ArgumentNullException.ThrowIfNull(query);
+
+        // Update he base properties first before calling handler, to ensure we update query source when applicable.
+        if (data != null)
+        {
+            UpdateBaseProperties(data, query);
+
+            // During the update, we allow changing the source.
+            var sourceName = data[nameof(Query.Source)]?.GetValue<string>();
+
+            if (!string.IsNullOrEmpty(sourceName) && query.Source != sourceName)
+            {
+                var querySource = _serviceProvider.GetKeyedService<IQuerySource>(sourceName);
+
+                if (querySource != null)
+                {
+                    query.Source = querySource.Name;
+                }
+            }
+        }
+
+        var updatingContext = new UpdatingQueryContext(query, data);
+        await _queryHandlers.InvokeAsync((handler, context) => handler.UpdatingAsync(context), updatingContext, _logger);
+
+        await _session.SaveAsync(query);
+        await _session.SaveChangesAsync();
```

Comment:
Why is calling `SaveChangesAsync()` everywhere necessary?

Reply:
just to ensure we generate new cache key after the commit took place.

Reply:
`SaveChangesAsync()` commits the changes to the DB, what I don't think any service should do, since it's ultimately done by [`IDocumentStore` for the whole scope](https://github.com/OrchardCMS/OrchardCore/blob/6312841ef0893e1585e93e325289794de4aada94/src/OrchardCore/OrchardCore.Data.YesSql/OrchardCoreBuilderExtensions.cs#L163). Committing early opens up data consistency issues where DB changes up until this point are persisted, then an exception is thrown, and others aren't, while due to the exception, all the changes should've been canceled. Since `DefaultQueryManager` has no control over what else is using the `ISession`, what gets committed to the DB in such a case may be nonsense. (With the classic transaction example of a bank: some code before this can deduct money from an account, and some code after it debit it to another one, but `SaveChangesAsync()` can cause only the first to be persisted.)

I think you can just regenerate the identifier on each change. On the off chance that the transaction is rolled back in the end, at most some GraphQL clients will do an unnecessary lookup, but we can't get much worse data inconsistency.

Reply:
I agree with you. I am actually thinking we should keep using DocumentManager for this since we need to cache query names and we need to manage the id for the schema builder. It may better than splitting everything into a separate document. However everything else this PR presents will stay.

Reply:
This is no longer an issue since we now use a new single document for this. So we no longer need to call SaveChangesAsync.